repose-aggregator/components/filters/keystone-v2/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2Filter.scala

```diff
@@ -0,0 +1,360 @@
+                      if (groups.isEmpty) {
+                        Pass(headers)
+                      } else {
+                        val groupsHeader = PowerApiHeader.GROUPS.toString -> groups.mkString(",")
+                        Pass(headers + groupsHeader)
+                      }
+                    case Some(Failure(e)) =>
+                      logger.error(s"Could not get groups: ${e.getMessage}")
+                      Pass(headers)
+                    case None => Pass(headers)
+                  }
+                case reject: Reject => reject
+              }
+
+              val addHeaders = userGroups match {
+                case Pass(headers) =>
+                  val userHeaders = Map(PowerApiHeader.USER.toString -> validToken.username,
+                    OpenStackServiceHeader.USER_NAME.toString -> validToken.username,
+                    OpenStackServiceHeader.USER_ID.toString -> validToken.userId)
+
+                  val xAuthHeader = uriTenantOption match {
+                    case Some(uriTenant) =>
+                      OpenStackServiceHeader.EXTENDED_AUTHORIZATION.toString -> s"$X_AUTH_PROXY $uriTenant"
+                    case None =>
+                      OpenStackServiceHeader.EXTENDED_AUTHORIZATION.toString -> X_AUTH_PROXY
+                  }
+
+                  val rolesHeader = if (config.getIdentityService.isSetRolesInHeader) {
+                    Map(OpenStackServiceHeader.ROLES.toString -> validToken.roles.mkString(","))
+                  } else {
+                    Map.empty[String, String]
+                  }
+
+                  val tenantName = OpenStackServiceHeader.TENANT_NAME.toString -> validToken.tenantName
+
+                  val defaultRegion = validToken.defaultRegion.map(dr => OpenStackServiceHeader.DEFAULT_REGION.toString -> dr)
+
+                  val contactId = validToken.contactId.map(ci => OpenStackServiceHeader.CONTACT_ID.toString -> ci)
+
+                  val expirationDate = OpenStackServiceHeader.X_EXPIRATION.toString -> validToken.expirationDate
+
+                  val impersonatorId = validToken.impersonatorId.map(iid => OpenStackServiceHeader.IMPERSONATOR_ID.toString -> iid)
+                  val impersonatorName = validToken.impersonatorName.map(in => OpenStackServiceHeader.IMPERSONATOR_NAME.toString -> in)
+
+                  val identityStatus = OpenStackServiceHeader.IDENTITY_STATUS.toString -> IdentityStatus.Confirmed.toString
+
+                  Pass(headers ++ userHeaders ++ rolesHeader + tenantName + xAuthHeader ++ defaultRegion ++ contactId
+                    + expirationDate ++ impersonatorId ++ impersonatorName + identityStatus)
+                case reject: Reject => reject
+              }
+
+              addHeaders
+            case Failure(x: IdentityAdminTokenException) =>
+              Reject(SC_INTERNAL_SERVER_ERROR, failure = Some(x))
+            case Failure(x: IdentityCommuncationException) =>
```

Comment:
I assume this is the case for getting rate limited by identity, how are you going to populate the header for when they can try again. Actually bad gateway is the wrong one then, it should service unavailable, so you have another case here.